websocket branch code review tweaks

oxtoacart opened the pull request:
@atavism I'm providing feedback and tweaks for the websocket branch code review here.  It's not ready yet, but I'll add code and comments as I proceed with my review and let you know when it's ready.

oxtoacart committed:
Added default cloud proxiedsites to binary

oxtoacart commented on src/github.com/getlantern/flashlight/client/config.go:
```diff
@@ -18,9 +18,8 @@ type ClientConfig struct {
 	DumpHeaders    bool // whether or not to dump headers of requests and responses
 	FrontedServers []*FrontedServerInfo
 	ChainedServers map[string]*ChainedServerInfo
+	ProxiedSites   *proxiedsites.Config // List of proxied site domains that get routed through Lantern rather than accessed directly
```
I moved this so that ProxiedSites comes before MasqueradeSets in the yaml.  It's more interesting and more likely to be edited by the user, so this is a nice convenience.

oxtoacart commented on src/github.com/getlantern/flashlight/config/config.go:
```diff
@@ -228,6 +228,11 @@ func (cfg *Config) applyClientDefaults() {
 	}
 
+	if cfg.Client.ProxiedSites.Cloud == nil || len(cfg.Client.ProxiedSites.Cloud) == 0 {
+		log.Debugf("Loading default cloud proxiedsites")
```
I'm now defaulting the list of cloud proxiedsites based on a list kept in the binary itself.

oxtoacart replied:
This is just to ensure the best possible out-of-the-box experience.

oxtoacart commented on src/github.com/getlantern/flashlight/genconfig/genconfig.go:
```diff
@@ -90,6 +88,11 @@ func main() {
 	if err != nil {
 		log.Fatalf("Unable to format masquerades.go: %s", err)
 	}
+	generateTemplate(model, proxiedSitesTmpl, "../config/proxiedsites.go")
```
Generates proxiedsites.go with default list.

oxtoacart commented on src/github.com/getlantern/proxiedsites/proxiedsites.go:
```diff
@@ -33,12 +33,12 @@ var (
 	Deletions []string `json:"Deletions, omitempty"`
+
+	// Global list of white-listed domains
+	Cloud []string `json:"-"`
```
Users are more likely to want to change the Additions and Deletions in the yaml, so I moved Cloud below them to have it appear further down in the YAML.

oxtoacart committed:
Refactoring proxiedsites

Commit:
fix open external url

oxtoacart committed:
Refactoring flashlight ui

oxtoacart committed:
Merge commit 'aa7a81a81556a407b820db66c6614a350a876c61' as 'src/github.com/gorilla/mux'

oxtoacart committed:
Squashed 'src/github.com/gorilla/mux/' content from commit 8a875a0

git-subtree-dir: src/github.com/gorilla/mux
git-subtree-split: 8a875a034c69b940914d83ea03d3f1299b4d094b

oxtoacart committed:
Merge commit 'ad17881dfe23c2d23e363cffd156fc03c6cb9a3e' as 'src/github.com/gorilla/context'

oxtoacart committed:
Squashed 'src/github.com/gorilla/context/' content from commit 215affd

git-subtree-dir: src/github.com/gorilla/context
git-subtree-split: 215affda49addc4c8ef7e2534915df2c8c35c6cd

oxtoacart committed:
[gost] Added github.com/gorilla/mux and its dependencies

oxtoacart committed:
Refactoring flashlight ui

oxtoacart committed:
Refactoring flashlight ui

Commit:
remove settings button and contact link

atavism commented:
@oxtoacart I know you're in the middle of reviewing changes, but I found a couple things to tweak in the proxied sites implementation (two new error messages + some additional tests I already wrote). I just committed it to this branch.

Commit:
add missing unit tests

Commit:
add missing unit tests

oxtoacart committed:
Refactoring flashlight ui

oxtoacart committed:
Merged

oxtoacart committed:
Removed dead code

oxtoacart committed:
Renamed http.go to ui.go

oxtoacart commented:
@xiam This still needs to be reviewed, but I think the API won't change very much from this point, so if you want to use this for your geolocation stuff, it's ready.

@atavism This is ready for review.  I refactored a bunch of stuff to achieve a cleaner separation between proxiedsites and flashlight, and within flashlight to achieve a cleaner separation between the general UI code and the proxiedsites-specific functionality.  I also built the default proxied sites list into the binary.

In testing, I see two problems that I'm pretty sure I didn't introduce:
1. When running with the default proxied sites list of 7000 or so sites, making any changes in the textarea is dog slow.  That's because the validation is checking every line on every keystroke.  This is unusable as it stands.
2. When testing with two windows, updates from one window will show up in the other window a few times, but then they stop applying.  I tried fixing this using $apply, which I think helped a little, but it's still inconsistent.

oxtoacart commented on src/github.com/getlantern/flashlight/client/client.go:
```diff
@@ -78,4 +78,7 @@ func (client *Client) Configure(cfg *ClientConfig) {
 	client.initReverseProxy(bal, cfg.DumpHeaders)
+	client.priorCfg = cfg
+	client.priorTrustedCAs = &x509.CertPool{}
+	*client.priorTrustedCAs = *globals.TrustedCAs
```
@atavism This is something that was broken from before your changes.  I noticed that we never stored the old client configuration, so everytime any part of the system configuration changed, the client would reconfigure itself.  This fixes that.

oxtoacart commented on genassets.bash:
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 
-tarfs -pkg http src/github.com/getlantern/ui/app > src/github.com/getlantern/flashlight/http/resources.go
+tarfs -pkg ui src/github.com/getlantern/ui/app > src/github.com/getlantern/flashlight/ui/resources.go
```
I renamed http to ui.

oxtoacart commented on src/github.com/getlantern/flashlight/client/config.go:
```diff
@@ -19,8 +18,6 @@ type ClientConfig struct {
 	ChainedServers map[string]*ChainedServerInfo
 	MasqueradeSets map[string][]*fronted.Masquerade
-
-	ProxiedSites *proxiedsites.Config // List of proxied site domains that get routed through Lantern rather than accessed directly
```
I moved proxiedsites.Config from client into the base config.  This is to keep the client from reconfiguring itself needlessly, since it actually doesn't care about the proxied sites.

oxtoacart commented on src/github.com/getlantern/flashlight/config/config.go:
```diff
@@ -205,6 +210,22 @@ func (cfg *Config) ApplyDefaults() {
+
+	if cfg.ProxiedSites.Cloud == nil || len(cfg.ProxiedSites.Cloud) == 0 {
+		log.Debugf("Loading default cloud proxiedsites")
+		cfg.ProxiedSites.Cloud = defaultProxiedSites
```
Here is where we default the proxiedsites list to the default list (generated by genconfig).

oxtoacart commented on src/github.com/getlantern/flashlight/proxiedsites/proxiedsites.go:
```diff
@@ -0,0 +1,70 @@
+
+func start() {
+	// Register the PAC handler
+	url := ui.Handle("/proxy_on.pac", http.HandlerFunc(proxiedsites.ServePAC))
```
I didn't see any reason to store the pac file on disk, and it added a fair amount of complexity, so now the proxiedsites package just serves up the PAC directly.

oxtoacart commented on src/github.com/getlantern/flashlight/proxiedsites/proxiedsites.go:
```diff
@@ -0,0 +1,70 @@
+	log.Debugf("Serving PAC file at %v", url)
+
+	// Establish a channel to the UI for sending and receiving updates
+	uichannel = ui.NewChannel("/data", func(write func([]byte) error) error {
```
@atavism @xiam UIChannel is an abstraction that I added to represent a channel of communication to the UI.  All browser windows are represented by this channel, so that when any of them write, we get the message, and if we write to the channel, all of the browsers get the message.  I use this abstraction here for the proxiedsites functionality, and @xiam can use it elsewhere for the geoip functionality (just use a different path than `/data`).

xiam replied:
Got it. I'll need to send a message to `Out`, right?

oxtoacart replied:
Yes, like [here](https://github.com/getlantern/flashlight-build/pull/62/files#diff-c117cc8455c34de273b2b92fc7c9d133R33).

oxtoacart commented on src/github.com/getlantern/proxiedsites/proxiedsites.go:
```diff
@@ -1,269 +1,217 @@
+// sites (cloud) and user-applied customizations to that list. It provides an
+// implementation of the http.Handler interface that serves up a PAC file based
+// on the currently active proxied sites (cloud + additions - deletions).
 package proxiedsites
```
proxiedsites was simplified considerably.
1. It's all just package level functions now (no ProxiedSites data structure)
2. There's no pac file, it's just bytes in memory. ProxiedSites provides a function that can be used as an HTTP handler that serves up the pac file from memory.
3. Additions and Deletions were factored into a Delta data structure since these are sometimes sent independently of the Global list.
4. It follows the usual Configure() paradigm.  Also, by having Configure() return a Delta representing the changes, we no longer need any channels in this package.

atavism replied:
Where is the flashlight.yaml file now? Is that no longer being generated? I use to be able to update the list of proxied sites directly from there. How do I make on-the-fly changes to the list now without going through the UI?

oxtoacart replied:
It's in ~/Library/Application Support/Lantern/lantern.yaml

oxtoacart replied:
BTW, I try to be good about logging stuff like this, so if you're ever looking to see where stuff is coming from, the log output is useful:

```
flashlight: Placing logs in /Users/ox.to.a.cart/Library/Logs/Lantern
Feb 20 17:56:27.919 - flashlight: ---- flashlight version development (now) ----
Feb 20 17:56:27.919 - flashlight.config: Placing configuration in /Users/ox.to.a.cart/Library/Application Support/Lantern
```

oxtoacart commented on src/github.com/getlantern/ui/app/js/app.js:
```diff
@@ -73,25 +73,35 @@ var app = angular.module('app', [
+            if (msg.Additions) {
+              entries = _.union(entries, msg.Additions);
+            }
+            if (msg.Deletions) {
```
The original code was not applying deletions.  Now it does.

oxtoacart commented on src/github.com/getlantern/ui/app/js/app.js:
```diff
@@ -73,25 +73,35 @@ var app = angular.module('app', [
+            if (msg.Deletions) {
+              entries = _.difference(entries, msg.Deletions)
+            }
+            entries = _.compact(entries);
```
This removes blank lines.

oxtoacart commented on src/github.com/getlantern/yamlconf/yamlconf.go:
```diff
@@ -211,7 +211,7 @@ func (m *Manager) processUpdates() {
 		case delta := <-m.deltasCh:
 			log.Trace("Apply delta")
-			updated, err := m.copy(m.cfg)
+			updated, err := m.copy(m.getCfg())
```
This was a little fix for a potential concurrency issue on accessing m.cfg.

oxtoacart commented on src/github.com/gorilla/context/.travis.yml:
```diff
@@ -0,0 +1,9 @@
+language: go
```
I vendored gorilla.

oxtoacart commented on src/github.com/robertkrimen/otto/.gitignore:
```diff
@@ -1,5 +0,0 @@
-/.test
```
We don't need otto anymore since we're not parsing a PAC file.

atavism replied:
Right. I was just using it previously to test the validity of the updated PAC file.

oxtoacart replied:
Ah, my bad.  I saw it in proxiedsites.go rather than proxiedsites_test.go and assumed it was part of the production code.

oxtoacart replied:
For future reference, it's nice to put testing code in the _test.go files, which keeps it from getting compiled into the main app.

oxtoacart committed:
Removed set.v0

oxtoacart committed:
Vendored set.v0

oxtoacart committed:
Removed unused utility functions

oxtoacart committed:
Comment updates

oxtoacart committed:
Error handling fix

oxtoacart committed:
Removed checked in executables

oxtoacart committed:
Always show UI if uiaddr is set

oxtoacart committed:
Removing resources.go from Git

oxtoacart committed:
Removed big resources.go from Git

oxtoacart committed:
Optimized embedded of resources with tarfs

xiam commented:
Thank you @oxtoacart, I'm working now on the websocket-ox-tweaks branch.

xiam commented on src/github.com/getlantern/flashlight/ui/ws.go:
```diff
@@ -0,0 +1,153 @@
+
+func newUIChannel(url string) *UIChannel {
+	in := make(chan []byte, 100)
+	out := make(chan []byte)
```
I noticed that the out channel is not buffered (`newUIChannel()`), so a goroutine that writes to it would block until the message is received on the client by WriteMessage. Since we're already thinking on a situation when two or more browsers are opened it would probably be better to have a small buffer here.

oxtoacart replied:
Maybe.  Keep in mind that the network layer is buffering stuff too, so this isn't actually blocking until the message has been received by the client, it's just blocking until the data has been written to the network buffers.

xiam commented:
It looks like we're assuming we'll send only one kind of message through the proxy (data about proxied sites): https://github.com/getlantern/flashlight-build/blob/websocket-ox-tweaks/src/github.com/getlantern/ui/app/js/app.js#L80

I'm I understanding that right?

So right now I can't use the same onMessage to receive a message with the user's geolocation data.

To fix that, I'm thinking on creating a simple envelope like:

``` json
{
 "MessageType": "ProxiedSites",
 "Data": { /* original data */ },
}
```

I would use another MessageType for the user's geolocation information and handle it in the same onMessage.

And in order to do that, I just have to add an `Envelope` struct to the `ws.go` file and update a few lines in `proxiedsites.go` and in `app.js`.

Do you have another suggestion?

atavism commented:
> And in order to do that, I just have to add an Envelope struct to the ws.go file and update a few lines in proxiedsites.go and in app.js.

Data was intended to support multiple message types anyway so that sounds like a good change to make.

oxtoacart commented:
@xiam Actually the whole reason that I introduced the UIChannel abstraction is so that you can set up a separate channel for things other than proxiedsites.  My intent was not for you to use the same UIChannel as proxiedsites.

atavism commented:
@oxtoacart Who did you want to review these changes?

xiam commented:
@oxtoacart What if I leave /data as a reserved websocket for proxiedsites and create a new one to handle miscellaneous messages? The user's geolocation information could be the first of those messages and we'd leave the implementation ready for any future usage, such as reporting errors, etc.

In the other hand, implementing everything in `/data` would have the additional benefit of having only one websocket connection opened (just one error management routine and one point of failure less to worry about).

atavism commented:
@oxtoacart Are changes to the UI code no longer automatically detected? It looks like I have to restart flashlight each time now.

oxtoacart commented:
> Who did you want to review these changes?

@atavism You.  Just assigned it to you.

oxtoacart commented:
> Are changes to the UI code no longer automatically picked up? It looks like I have to restart flashlight each time now.

How are you running flashlight?

oxtoacart commented on src/github.com/getlantern/flashlight/ui/ui.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+const (
+	localResourcesPath = "../ui/app"
```
@atavism UI resources are specified using a relative path that assumes that you're running flashlight from within the flashlight folder during development.

atavism replied:
I've been running flashlight from the flashlight-build directory. So this is no longer looking for a ui directory in `src/github.com/getlantern`? Running flashlight from `src/github.com/getlantern/flashlight` I'm seeing 

> go run flashlight.go -role client -addr localhost:10080 -uiaddr localhost:16785 -cloudconfig http://s3.amazonaws.com/lantern_config/cloud.valencia.yaml.gz
> ./flashlight.go:228: undefined: Asset

myleshorton replied:
@atavism for whatever reason I see the same thing when running with `go run` but not with the precompiled ./flashlight. So this works for me:

```
./flashlight -role client -addr localhost:10080 -uiaddr localhost:16785 -cloudconfig http://s3.amazonaws.com/lantern_config/cloud.valencia.yaml.gz
```

oxtoacart replied:
@myleshorton Just changed this to make it relative to the GOPATH so it should work no matter where you run flashlight.  Can you please try again?

myleshorton replied:
I think this was intended for @atavism

On Fri, Feb 20, 2015 at 10:22 AM, oxtoacart notifications@github.com
wrote:

> In src/github.com/getlantern/flashlight/ui/ui.go
> https://github.com/getlantern/flashlight-build/pull/62#discussion_r25090293
> :
> 
> > +package ui
> > +
> > +import (
> > -   "fmt"
> > -   "net"
> > -   "net/http"
> > -   "path"
> > -   "path/filepath"
> >   +
> > -   "github.com/getlantern/golog"
> > -   "github.com/getlantern/tarfs"
> > -   "github.com/skratchdot/open-golang/open"
> >   +)
> >   +
> >   +const (
> > -   localResourcesPath = "../ui/app"
> 
> @myleshorton https://github.com/myleshorton Just changed this to make
> it relative to the GOPATH so it should work no matter where you run
> flashlight. Can you please try again?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/getlantern/flashlight-build/pull/62/files#r25090293.

## 

## 

President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

oxtoacart commented:
@xiam 

> What if I leave /data as a reserved websocket for proxiedsites and create a new one to handle miscellaneous messages? The user's geolocation information could be the first of those messages and we'd leave the implementation ready for any future usage, such as reporting errors, etc.

Maybe.  It'll be clearer once we have requirements for other data.

> In the other hand, implementing everything in /data would have the additional benefit of having only one websocket connection opened (just one error management routine and one point of failure less to worry about).

These are websockets to a local server, so I wouldn't worry about failures.  In terms of error handling, the only handling that we do is reconnecting.  That can be gotten for free by using a library like this one - https://github.com/joewalnes/reconnecting-websocket

Also, if we ever have application-specific error handling, it will probably vary by channel.  So for example, if we can't get your authentication information, that might cause us to actually show an error, whereas if we can't get the latest stats for peers we may just ignore that.

In general, I like the pattern of dedicated channels because it avoids having giant `switch (messageType)` statements everywhere.  It's possible to build a nice API like that and multiplex messages over a single websocket, but I don't know that there's much benefit to that.  And the benefit to giving everything its own websocket is that a slow reader in one subsystem doesn't hold up processing in another.

Just my 2 cents.

atavism commented:
> In general, I like the pattern of dedicated channels because it avoids having giant switch (messageType) statements everywhere. It's possible to build a nice API like that and multiplex messages over a single websocket, but I don't know that there's much benefit to that. And the benefit to giving everything its own websocket is that a slow reader in one subsystem doesn't hold up processing in another.

We didn't really get to discuss the merits of both approaches at all. My original implementation used a single websocket for the reasons already outlined. Nothing about the previous API necessitated using a giant switch statement: we could have configured the mux router to use multiple HTTP handlers with a single websocket. The issue of a slow reader could have been handled across the board by adding a read timeout to the processing time of every request. There was some error handling already that's since been removed.

oxtoacart committed:
Made local resources relative to GOPATH

oxtoacart commented:
@atavism 

> I've been running flashlight from the flashlight-build directory. So this is no longer looking for a ui directory in `src/github.com/getlantern`?

Yeah, sorry, I changed that. I've updated it now to make the path relative to the GOPATH so it should work no matter where you run flashlight.

> Running flashlight from `src/github.com/getlantern/flashlight` I'm seeing 
> 
> > go run flashlight.go -role client -addr localhost:10080 -uiaddr localhost:16785 -cloudconfig http://s3.amazonaws.com/lantern_config/cloud.valencia.yaml.gz
> > ./flashlight.go:228: undefined: Asset

Very strange.  Do you have the file icons.go in the same folder as flashlight.go?

atavism commented:
> Very strange. Do you have the file icons.go in the same folder as flashlight.go?

Yes.

oxtoacart commented:
> We didn't really get to discuss the merits of both approaches at all.

We're discussing it now.

> My original implementation used a single websocket for the reasons already outlined. Nothing about the previous API necessitated using a giant switch statement: we could have configured the mux router to use multiple HTTP handlers with a single websocket.

Can you show me how please (code sample, reference link)?  If we can make this work, we should in theory be able to keep the UIChannel API on the server and just use a single websocket.

> The issue of a slow reader could have been handled across the board by adding a read timeout to the processing time of every request.

Not really.  What I'm talking about is that if we have multiple consumers of messages from the websocket, the read loop will be dispatching to those consumers.  If one of those consumers is slow in doing its work, it will hold up subsequent reads.  We could introduce additional channels for each consumer to mitigate this, but that adds complexity.

> There was some error handling already that's since been removed.

Sorry about that.  Can you point me to the original code please?

oxtoacart commented:
@atavism 

> > Very strange. Do you have the file icons.go in the same folder as flashlight.go?
> 
> Yes.

Well, that's where the Asset function is defined.  Can you please check if your icons.go has the Asset function?  You can also try doing a full rebuild with `go install -a` to see if that clears things up.

BTW, can you please copy and paste the value of your `$GOPATH`?

atavism commented:
```
➜  flashlight git:(websocket-ox-tweaks) ✗ go run flashlight.go -role client -addr localhost:10080 -uiaddr localhost:16785 -cloudconfig http://s3.amazonaws.com/lantern_config/cloud.valencia.yaml.gz
# command-line-arguments
./flashlight.go:228: undefined: Asset
➜  flashlight git:(websocket-ox-tweaks) ✗ echo $GOPATH
/Users/todd/Dev/go/src/github.com/getlantern/flashlight-build
```

In icons.go:

```
   // Asset loads and returns the asset for the given name.
   // It returns an error if the asset could not be found or
   // could not be loaded.
   func Asset(name string) ([]byte, error) {
       cannonicalName := strings.Replace(name, "\\", "/", -1)
       if f, ok := _bindata[cannonicalName]; ok {
           return f()
       }
       return nil, fmt.Errorf("Asset %s not found", name)
   }
```

oxtoacart commented:
`go run flashlight.go`

That ignores all the other go files.  Try this:

`go install && flashlight`

myleshorton commented:
> go run flashlight.go
> 
> That ignores all the other go files. Try this:
> 
> go install && flashlight

Oh right -- that's what I was saying worked for me (well, go build && ./flashlight), but I didn't fully get why. I thought go run xxx.go would pull in dependencies and build them, but I guess not. This is clear with:

```
$ flashlight (websocket-ox-tweaks)]$ go build flashlight.go
# command-line-arguments
./flashlight.go:228: undefined: Asset
```

atavism commented on src/github.com/getlantern/flashlight/ui/ui.go:
```diff
@@ -0,0 +1,72 @@
+var (
+	log = golog.LoggerFor("ui")
+
+	localResourcesPath = filepath.Join(os.Getenv("GOPATH"), "src/github.com/getlantern/ui/app")
```
The code I had before actually looked for the ui directory relative to the source file of the function being invoked which conveniently handled most use cases. It used [runtme.Caller](http://golang.org/pkg/runtime/#Caller)

oxtoacart replied:
Ahh, cool.  Sorry about that - restored now.  I think I missed it because I ditched the branch that bypassed tarfs.

atavism commented on src/github.com/getlantern/flashlight/ui/ws.go:
```diff
@@ -0,0 +1,153 @@
+		log.Tracef("Closing all websockets to %v", c.URL)
+		c.m.Lock()
+		for _, conn := range c.conns {
+			conn.ws.Close()
```
What about deleting the connection ID from the map?

atavism replied:
Oh, this closes every connection. Nevermind.

oxtoacart replied:
You're right, I should delete them.  Fixed.  Thanks!

atavism commented on src/github.com/getlantern/proxiedsites/proxiedsites.go:
```diff
@@ -1,269 +1,218 @@
-	go ps.updatePacFile()
-	return ps
+	sort.Strings(l)
+	return l
```
Why use this custom implementation over the existing set.StringSlice? Because it sorts at the end? It would help to at least mention that in the comment.

oxtoacart replied:
Oh hey cool, that saves some code.  Fixed!

atavism commented on src/github.com/getlantern/proxiedsites/proxiedsites.go:
```diff
@@ -1,269 +1,218 @@
+	resp.WriteHeader(http.StatusOK)
+	cfgMutex.RLock()
+	resp.Write([]byte(pacFile))
+	cfgMutex.RUnlock()
```
Nice. Definitely makes sense to store it in-memory and skip saving a file entirely since we already write the latest changes to the YAML configuration.

oxtoacart replied:
Thanks.

oxtoacart committed:
Restored logic for figuring out local resources directory

oxtoacart committed:
Code review fixes

oxtoacart committed:
Code review fixes

atavism commented on src/github.com/getlantern/flashlight/ui/ws.go:
```diff
@@ -0,0 +1,153 @@
+			c:  c,
+			ws: ws,
+		}
+		c.conns[conn.id] = conn
```
What purpose is the connection ID actually serving here? You just ignore the ID everywhere you iterate through the connections anyway.

oxtoacart replied:
I use it to delete connections from the map of connections upon closing them.

atavism commented on src/github.com/getlantern/flashlight/ui/ws.go:
```diff
@@ -0,0 +1,153 @@
+		_, b, err := c.ws.ReadMessage()
+		if err != nil {
+			if err != io.EOF {
+				log.Debugf("Error reading from UI: %v", err)
```
Not really a Debug message.

oxtoacart replied:
Fixed.

oxtoacart committed:
Code review fixes

atavism commented on src/github.com/getlantern/flashlight/ui/ui.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}()
+	uiaddr = fmt.Sprintf("http://%v", l.Addr().String())
+	log.Debugf("UI available at %v", uiaddr)
```
Do we want to continue reporting the UI is available if Serve() fails? I'd just move that debugging info after the nil error check (and return) inside the goroutine.

oxtoacart replied:
I can't do that because Serve blocks until the server is done serving.

oxtoacart commented:
@atavism Thanks for reviewing!  I hope that the modified structure makes sense and that I wasn't too gratuitous in my tweaks.
